src/httpclient.c

```diff
@@ -224,23 +243,24 @@ static void tunnel_on_udp_sock_read(h2o_socket_t *sock, const char *err)
 
     /* send the datagram directly or encapsulated on the stream */
     if (udp_write != NULL) {
-        h2o_iovec_t datagram = h2o_iovec_init(buf, rret);
+        h2o_iovec_t datagram = h2o_iovec_init(buf, rret + context_id_len);
         udp_write(client, &datagram, 1);
     } else {
         /* append UDP chunk to the input buffer of stdin read socket! */
         uint8_t header[3] = {0}, *header_end = quicly_encodev(header + 1, (uint64_t)rret);
         h2o_buffer_append(&std_in.sock->input, header, header_end - header);
-        h2o_buffer_append(&std_in.sock->input, buf, rret);
+        h2o_buffer_append(&std_in.sock->input, buf, rret + context_id_len);
         /* pretend as if we read from stdin */
         stdin_on_read(std_in.sock, NULL);
     }
 }
 
 static void tunnel_on_udp_read(h2o_httpclient_t *client, h2o_iovec_t *datagrams, size_t num_datagrams)
 {
+    size_t context_id_len = (connect_udp_is_draft03) ? 0 : 1;
     for (size_t i = 0; i != num_datagrams; ++i) {
         const h2o_iovec_t *src = datagrams + i;
-        struct iovec vec = {.iov_base = src->base, .iov_len = src->len};
+        struct iovec vec = {.iov_base = src->base + context_id_len, .iov_len = src->len - context_id_len};
```

Comment:
IIUC, `datagrams` is an array of "UDP Proxying HTTP Datagram Payload" as defined in [RFC 9298 Section 5](https://www.rfc-editor.org/rfc/rfc9298.html#name-http-datagram-payload-forma). Assuming that, I think we need to parse and check the context ID before handling it as a UDP datagram?

Reply:
See if this change addresses your comment.